common/src/main/java/uk/gov/gchq/palisade/io/BytesSuppliedInputStream.java

```diff
@@ -83,11 +95,56 @@ public int read() throws IOException {
             return -1;
         }
 
-        byte b = bytes[i];
+        byte b = bytes[index];
         if (null != bytes && LOGGER.isDebugEnabled()) {
             LOGGER.debug("Reading byte {}", new String(new byte[]{b}));
         }
-        i++;
+        index++;
         return b & 0xff;
     }
+
+    private void resetOnNullNewBytes() {
+        bytes = null;
+        bytesCount = 0;
+        index = 0;
+    }
+
+    @Override
+    public int read(final byte[] b, final int off, final int len) throws IOException {
+        if (len == 0) {
+            return 0;
+        }
+        if (end) {
+            return -1;
+        }
+        if (null == bytes || index >= bytesCount) {
+            final Bytes newBytes = supplier.get();
+            index = 0;
+            if (null == newBytes) {
+                resetOnNullNewBytes();
+            } else if (((bytes = newBytes.getBytes()) != null) && ((bytesCount = newBytes.getCount()) > bytes.length)) {
+                bytesCount = bytes.length;
+            }
+        }
+        if (null == bytes || bytesCount == 0) {
+            LOGGER.debug("Reached the end of the buffer");
+            end = true;
+            return -1;
+        }
+
+        int copyAmount = len;
+        int positionMinus1 = Arrays.binarySearch(bytes, 0, bytes.length, (byte) -1); //find end point in the stream
+        if ((positionMinus1 >= 0) && (positionMinus1 < copyAmount)) {
+            copyAmount = positionMinus1;
+        }
+        if ((bytesCount - index) < copyAmount) {
+            copyAmount = bytesCount - index;
+        }
+        if (copyAmount <= 0) {
+            return -1;
+        }
+        System.arraycopy(bytes, index, b, off, copyAmount);
+        index += copyAmount;
+        return copyAmount;
+    }
```

Comment:
Unfortunately, this piece of code from line 135-149 has several issues that will need to be addressed before I can grant merge approval.

Binary search can not be used in a read method for the following reasons:
 - It only works on sorted arrays. It can't be used on arbitrary byte streams.
 - Ignoring the above, it is searching the entire byte array each time, this would lead to incorrect behaviour where index is greater > 0
 - Crucially, there is no guarantee that -1 will be in the byte array at all. It only indicates "end" when returned by Java from a read() method. Remember there may be legitimate "-1"s in the data stream, so searching for them cannot indicate "end of stream"
- It would be slow

The logic for a read method like this should something like:
Is there anything available in the buffer? If not then refill it.
Can the requested amount of bytes be re-filled from the existing buffer? If yes, do so and update index pointer.
If not, return what is available. This means the buffer will be completely used upon the next call to read()